src/TableRow.js

```diff
@@ -72,14 +78,58 @@ class TableRow extends React.Component {
     }
   }
 
-  shouldComponentUpdate(nextProps) {
-    return !!(this.props.visible || nextProps.visible);
+  shouldComponentUpdate({
+    expandedRowKeys,
+    scrolling,
+    prefixCls,
+    columns,
+    record,
+    rowKey,
+    index,
+    onRow,
+    indent,
+    indentSize,
+    hovered,
+    height,
+    virtualizedRelatedStyle,
+    visible,
+    components: {
+      body: { row, cell },
+    },
+    hasExpandIcon,
+    renderExpandIcon,
+    renderExpandIconCell,
+  }) {
+    const shouldNotUpdate =
+      scrolling === this.props.scrolling &&
+      prefixCls === this.props.prefixCls &&
+      isEqual(expandedRowKeys, this.props.expandedRowKeys) &&
+      isEqual(columns, this.props.columns) &&
+      shallowequal(record, this.props.record) &&
+      rowKey === this.props.rowKey &&
+      index === this.props.index &&
+      indent === this.props.indent &&
+      indentSize === this.props.indentSize &&
+      hovered === this.props.hovered &&
+      height === this.props.height &&
+      visible === this.props.visible &&
+      hasExpandIcon === this.props.hasExpandIcon &&
+      renderExpandIcon === this.props.renderExpandIcon &&
+      renderExpandIconCell === this.props.renderExpandIconCell &&
+      shallowequal(row, this.props.components.body.row) &&
+      shallowequal(cell, this.props.components.body.cell) &&
+      shallowequal(virtualizedRelatedStyle, this.props.virtualizedRelatedStyle) &&
+      shallowequal(onRow, this.props.onRow);
+    return !shouldNotUpdate;
```

Comment:
Too complicated, can we simplify it? A row which is invisible or not in viewport should not update.